### PR TITLE
Alerting: Move annotation functionality behind a history persistence interface

### PR DESCRIPTION
**Pull request by alexweav**
**What this PR does / why we need it**:

Our current annotation system can be thought of as an early form of state history.

This PR creates an interface, `state.Historian`, for the state history work. It then treats the annotation logic as an existing implementation of state history.

This gives us a hook where we can explore other historian implementations. It also simplifies dependencies of state manager + tests, so this feels reasonable to merge into main now.

**Which issue(s) this PR fixes**:

Contrib https://github.com/grafana/grafana/issues/52496

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/state/historian/annotation.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+)
+
+// AnnotationStateHistorian is an implementation of state.Historian that uses Grafana Annotations as the backing datastore.
+type AnnotationStateHistorian struct {
```
Comment (alexweav) [inline pkg/services/ngalert/state/historian/annotation.go:20]: I'm keeping in mind that the state history may not necessarily _replace_ annotations, this isn't really an explored problem yet.

Even if we keep annotations, we can still treat them as an implementation of this interface, because it's still effectively recording history - we can easily keep them around with a composite implementation.

This approach here does not mean we are getting rid of annotations.

File: pkg/services/ngalert/state/persist.go [hunk 2]
```diff
@@ -19,3 +22,8 @@ type InstanceStore interface {
 type RuleReader interface {
 	ListAlertRules(ctx context.Context, query *models.ListAlertRulesQuery) error
 }
+
+// Historian maintains an audit log of alert state history.
+type Historian interface {
+	RecordState(ctx context.Context, rule *ngmodels.AlertRule, labels data.Labels, evaluatedAt time.Time, currentData, previousData InstanceStateAndReason)
```
Comment (alexweav) [inline pkg/services/ngalert/state/persist.go:28]: The args to `RecordState` will definitely change in future PRs. I just kept the same args from the old `annotateState` helper for the time being. This was done to not over-complicate the diff of this PR.

I know the args aren't very elegant right now, the important bit is the abstraction and naming.

File: pkg/services/ngalert/state/historian/annotation.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashboards.DashboardService, log log.Logger) *AnnotationStateHistorian {
+	return &AnnotationStateHistorian{
+		annotations: annotations,
+		dashboards:  dashboards,
+		log:         log,
+	}
+}
+
+func (h *AnnotationStateHistorian) RecordState(ctx context.Context, rule *ngmodels.AlertRule, labels data.Labels, evaluatedAt time.Time, currentData, previousData state.InstanceStateAndReason) {
```
Comment (alexweav) [inline pkg/services/ngalert/state/historian/annotation.go:34]: Logically identical to `annotateState` from `manager.go`. Some vars have been renamed, that's it.

Comment (joeblubaugh) [reply pkg/services/ngalert/state/historian/annotation.go:34]: I think this should be able to handle multiple states in a single call - too many small transactions is an ongoing performance problem. Giving the implementation the control over whether and how to batch writes is important.

I think it'd be ok to do this in a follow-up PR but I also think it's very important.

**Review by joeblubaugh — COMMENTED** [review]
I don't know if this is the right direction to go with saving state history, because I don't think we're saving state history in the right part of alert state management right now.

We're currently doing two writes of nearly identical data four lines apart in `ProcessEvalResults`. This one is in `setNextState` and the other is in `saveState`. Those should be done at one time, in the same place. I'd like to see that situation straightened out before we call the current interface a meaningful name like "Historian".

When #55350 lands, we will have batched up the saveState calls for an individual rule. We should take that one step farther and absorb the state annotation calls as well. At that point I think there's a reasonable "historian" interface to be extracted.

File: pkg/services/ngalert/state/historian/annotation.go (hunk 3 above)
Comment (alexweav) [reply pkg/services/ngalert/state/historian/annotation.go:34]: Agree...eventually, but this is also meant to be a refactor.

The choice to make the args here identical to the old helper function was an intentional one. As a reviewer, I never like it when a chunk of code gets moved and the logic gets changed at the same time, it becomes hard to follow what exactly changed in the diff. So, I leaned towards changing the structure of moved code as little as possible within the same PR.

**Comment by alexweav** [comment]
In general, I'd like to note that this change is not trying to solve the world just yet. Things like write batching/transaction management, the structure of the `state` package, and the fact that our alert instance table has some significant redundancies with other writes are all _absolutely_ major considerations of this effort.

But, I don't feel that should preclude us from making incremental changes aimed in the right direction. Untangling these into distinct problems that aren't so tightly coupled to each other is a reasonable first step I feel.

I fully expect this interface will change further. In fact, I intentionally stopped myself from changing it further, in order to keep the diff of the moved code 1:1 with the original - to make this PR easier to follow for reviewers.

File: pkg/services/ngalert/state/historian/annotation.go (hunk 3 above)
Comment (alexweav) [reply pkg/services/ngalert/state/historian/annotation.go:34]: I'd also like this to work in terms of DTOs more clearly - to use the actual `state.State` structs, and not the piecemeal fields that it takes right now. But, that takes some improvement of the code in the State package, which I felt was out of scope for this PR - that's what I'm looking into next.

I feel this is actually required before we can think about taking multiple states in the same call. The current structure isn't very conducive to it at all.

Comment (joeblubaugh) [reply pkg/services/ngalert/state/historian/annotation.go:34]: I think if I had a better sense of what was coming next after this PR I'd be able to judge the direction better. Right now I'm concerned about emphasizing a difference between current state and history, and doing too much experimentation behind the Historian interface without fixing the state manager's bad practice of writing everything down twice.

**Review by joeblubaugh — APPROVED** [review]
> aimed in the right direction.

I'd like to understand what a better API for a historian interface will look like after a few more changes. Let's talk about that in real-time soon.

File: pkg/services/ngalert/state/historian/annotation.go (hunk 3 above)
Comment (alexweav) [reply pkg/services/ngalert/state/historian/annotation.go:34]: I dont feel that annotations are a stable enough backend to experiment with a solution to this problem just yet. The actual history implementation needs to iterate a bit, first.

Annotations, have their own eviction system that we do not control, many other things write to annotations, we have few guarantees about their durability.

There may be a point where we can do away with the instance table by calculating state based on the most recent transition in history. However, I don't think we'll be able to solve that well without a (rudimentary, even) state implementation that isn't based on annotations.

This is why I felt this abstraction layer needs to be added sooner rather than later. I want to try and reduce the coupling between these problems. I don't see this layer going away over the course of the project (maybe changing, sure).

**Comment by yuri-tceretian** [comment]
I think that before extracting this logic we should refactor the manager to submit annotations per rule in bulk, this will likely cause  a some intermediate structure to be created that will hold the current an previous state, and then that logic would be simply extracted.

**Comment by alexweav** [comment]
@yuri-tceretian I'm actively working on this, in a branch off this change
